[CI] StatelessReshardDisruptionIT testReshardWithDisruption failing

**Build Scans:**
- [elasticsearch-intake #40916 / part2](https://gradle-enterprise.elastic.co/s/o6jxoeubn7uu6)
- [elasticsearch-pull-request #151557 / part-2](https://gradle-enterprise.elastic.co/s/ttq6poaowdvww)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:stateless:internalClusterTest" --tests "org.elasticsearch.xpack.stateless.reshard.StatelessReshardDisruptionIT.testReshardWithDisruption" -Dtests.seed=2EDD5DD5EB848F71 -Dtests.locale=chr -Dtests.timezone=Etc/GMT-5 -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.stateless.reshard.StatelessReshardDisruptionIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testReshardWithDisruption'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.lang.IllegalStateException: Some shards are still open after the threadpool terminated. Something is leaking index readers or store references.
```

**Issue Reasons:**
- [main] 1 consecutive failures in test testReshardWithDisruption
- [main] 1 consecutive failures in step part2
- [main] 1 consecutive failures in pipeline elasticsearch-intake
- [main] 2 failures in test testReshardWithDisruption (0.3% fail rate in 599 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

I took a look at this and I think I found the root cause: 
  
The test fails at node close with a shard leak (what CI reports):

```
  testReshardWithDisruption {seed=[BDE46C0DF2D23F:E65F55E608AFAE19]} FAILED
      java.util.concurrent.ExecutionException: RuntimeException: IllegalStateException:
        Some shards are still open after the threadpool terminated. Something is leaking index readers or store references.
          at Node.awaitClose(Node.java:635)
          at InternalTestCluster.stopNode(InternalTestCluster.java:1807)
          at StatelessReshardDisruptionBaseIT.induceFailure(StatelessReshardDisruptionBaseIT.java:58)
          at StatelessReshardDisruptionIT.testReshardWithDisruption(StatelessReshardDisruptionIT.java:80)
```

  The real cause is a separate `AssertionError` on a background `stateless_shard_write` thread, queued and reported alongside it:

```
  WARNING: Uncaught exception in thread: Thread[node_t6][stateless_shard_write][T#1]
  java.lang.AssertionError: enqueue task failed: java.lang.IllegalStateException: Object store service is not running [CLOSED]
      at ObjectStoreService.enqueueTask(ObjectStoreService.java:744)
      at ObjectStoreService.uploadBatchedCompoundCommitFile(ObjectStoreService.java:683)
      at BatchedCompoundCommitUploadTask.uploadBatchedCompoundCommitFile(BatchedCompoundCommitUploadTask.java:187)
      at StatelessCommitService$ShardCommitState.runUploadWhenCommitIsReady(StatelessCommitService.java:2172)
      at BatchedCompoundCommitUploadTask.tryAction(BatchedCompoundCommitUploadTask.java:113)
      at RetryableAction$1.doRun(RetryableAction.java:112)
```

  And the two adjacent log lines showing the race (from one local reproduction):
```
  1> [23:54:33,139][INFO][ObjectStoreService][[pool-304-thread-1]] object store service closed   ← node_t6 service CLOSED
  2> WARNING: Uncaught exception in thread: Thread[node_t6][stateless_shard_write][T#1]            ← upload thread races it
  2> java.lang.AssertionError: enqueue task failed: ...IllegalStateException: ...not running [CLOSED]
```

  **What's happening:** an index node is stopped while a commit upload is in flight. ObjectStoreService is already CLOSED, so enqueueTask → ensureRunning() throws IllegalStateException. The old catch hit assert false, throwing an AssertionError (a Throwable, not an Exception) that escaped RetryableAction and skipped
  listener.onFailure(e). The listener never fired → its decRef() cleanup never ran → store ref leaked → Node.awaitClose failed.

  **Why it shows up in this IT**

 The bug is a race: a node must be uploading a commit at the exact moment it's shut down. This test makes that likely because resharding keeps commit uploads running almost constantly, while the disruption loop repeatedly stops index nodes — so a stop often coincides with an in-flight upload.

  **Reproduced in a couple of iterations - locally**

  At ~0.3–0.5% per iteration (matching CI's 0.3% over 599 runs), it reproduces roughly once per ~500 randomized iterations. Caught locally multiple times — seeds 7183166D869CFC3F (2 hits, REPLACE_FAILED_NODE), AB9B9CD35E2C4B69, BDE46C0DF2D23F; CI was 2EDD5DD5EB848F71 (RESTART). Every hit showed the identical
  signature: object store service closed immediately followed by enqueue task failed ... [CLOSED], then the leak at teardown.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -458,9 +458,6 @@ tests:
 - class: org.elasticsearch.indices.memory.breaker.HierarchyCircuitBreakerTelemetryIT
   method: testCircuitBreakerTripCountMetric
   issue: https://github.com/elastic/elasticsearch/issues/150700
-- class: org.elasticsearch.xpack.stateless.reshard.StatelessReshardDisruptionIT
-  method: testReshardWithDisruption
-  issue: https://github.com/elastic/elasticsearch/issues/150708
 - class: org.elasticsearch.xpack.esql.analysis.AnalyzerUnmappedGoldenTests
   method: testTypeConflictTimeseriesLongUnmappedWithCast
   issue: https://github.com/elastic/elasticsearch/issues/150722

--- a/x-pack/plugin/stateless/src/main/java/org/elasticsearch/xpack/stateless/objectstore/ObjectStoreService.java
+++ b/x-pack/plugin/stateless/src/main/java/org/elasticsearch/xpack/stateless/objectstore/ObjectStoreService.java
@@ -645,10 +645,14 @@ public class ObjectStoreService extends AbstractLifecycleComponent implements Cl
         }
     }
 
-    private void ensureRunning() {
+    private boolean isRunning() {
         final Lifecycle.State state = lifecycleState();
-        if (state == Lifecycle.State.INITIALIZED || state == Lifecycle.State.CLOSED) {
-            throw new IllegalStateException("Object store service is not running [" + state + ']');
+        return state != Lifecycle.State.INITIALIZED && state != Lifecycle.State.CLOSED;
+    }
+
+    private void ensureRunning() {
+        if (isRunning() == false) {
+            throw new IllegalStateException("Object store service is not running [" + lifecycleState() + ']');
         }
     }
 
@@ -741,7 +745,9 @@ public class ObjectStoreService extends AbstractLifecycleComponent implements Cl
                 }
             }
         } catch (Exception e) {
-            assert false : "enqueue task failed: " + e;
+            // Enqueueing can fail once the service is shutting down while uploads are still in flight: notify the listener so callers
+            // release the references they hold (e.g. commit refs) rather than leaking them.
+            assert isRunning() == false : "enqueue task failed while object store service is running: " + e;
             listener.onFailure(e);
         }
     }

--- a/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/objectstore/ObjectStoreServiceTests.java
+++ b/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/objectstore/ObjectStoreServiceTests.java
@@ -23,6 +23,7 @@ import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.action.ActionRequest;
 import org.elasticsearch.action.ActionResponse;
 import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.support.PlainActionFuture;
 import org.elasticsearch.blobcache.BlobCacheUtils;
 import org.elasticsearch.blobcache.shared.SharedBlobCacheService;
 import org.elasticsearch.blobcache.shared.SharedBytes;
@@ -191,6 +192,26 @@ public class ObjectStoreServiceTests extends ESTestCase {
         assertThat(settings.get(bucketName), equalTo(bucket));
         assertThat(settings.get("client"), equalTo(client));
         assertThat(settings.get("base_path"), equalTo(basePath));
+    }
+
+    /**
+     * When the object store service is closed while an upload is still in flight, enqueueing the upload must notify its listener
+     * rather than throwing, so the caller can release the references it holds (e.g. commit refs) instead of leaking them.
+     */
+    public void testEnqueueAfterCloseNotifiesListener() throws IOException {
+        try (var testHarness = new FakeStatelessNode(this::newEnvironment, this::newNodeEnvironment, xContentRegistry())) {
+            final var objectStoreService = testHarness.objectStoreService;
+
+            objectStoreService.close();
+            assertThat(objectStoreService.lifecycleState(), equalTo(Lifecycle.State.CLOSED));
+
+            final var future = new PlainActionFuture<Void>();
+            objectStoreService.uploadTranslogFile("translog-1", new BytesArray(new byte[] { 1, 2, 3 }), future);
+
+            // The listener must be failed (not stranded by a thrown error) so the caller can release its references.
+            final var e = expectThrows(IllegalStateException.class, future::actionGet);
+            assertThat(e.getMessage(), startsWith("Object store service is not running"));
+        }
     }
 
     public void testStartingShardRetrievesSegmentsFromOneCommit() throws IOException {

